ai/STA/Tactic/tactic.py

```diff
@@ -40,34 +29,37 @@ def __init__(self, game_state: GameState, player: Player, target: Pose=Pose(), a
         self.next_state = self.halt
         self.status_flag = Flags.INIT
         self.target = target
+        self.forbidden_areas = forbidden_areas if forbidden_areas is not None else []
 
     def halt(self) -> Idle:
-        """
-            S'exécute lorsque l'état courant est *Halt*. Générique pour arrêter n
-            'importe quelles tactiques enfants
-
-            :return: un nouvelle instance de l'action Idle pour le robot
-        """
         self.next_state = self.halt
         return Idle
 
     def exec(self) -> AICommand:
-        """
-            Exécute une *Action* selon l'état courant
-
-            :return: un AICommand
-        """
         next_ai_command = self.current_state()
         if not isinstance(next_ai_command, AICommand):
             raise RuntimeError("A tactic MUST return an AICommand, not a {}. {} is the culprit.".format(type(next_ai_command), self.current_state))
         self.current_state = self.next_state
+
+        if next_ai_command.target:
+            target_position = next_ai_command.target.position
+            target_to_position = Line(self.player.position, next_ai_command.target.position)
+            for area in self.forbidden_areas:
+                if target_position in area:
+                    intersections = area.intersect(target_to_position)
+                    if intersections:
+                        target_position = intersections[0]
+                    else:
+                        target_position = position_outside_area(target_position, area)
+                    new_target = Pose(target_position, next_ai_command.target.orientation)
+                    # This trailing _ is not for protected access, it was add to avoid a name conflict with the function replace ;)
+                    next_ai_command = next_ai_command._replace(target=new_target)
+                    break
+
```

Comment:
Could you put that inside a method, like `check_for_forbidden_area` or something?

Reply:
Yes